### PR TITLE
[CSI] Raw Block support

### PR DESCRIPTION
**Pull request by ggriffiths**
Signed-off-by: Grant Griffiths <grant@portworx.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

### CODE REVIEW

File: volume/drivers/fake/fake.go [hunk 1]
```diff
@@ -226,9 +226,6 @@ func (d *driver) Unmount(volumeID string, mountpath string, options map[string]s
 	if err != nil {
 		return err
 	}
-	if len(v.AttachPath) == 0 {
```
Comment (ggriffiths) [inline volume/drivers/fake/fake.go:229]: Needed for csisanity idempotency

File: csi/node.go [hunk 2]
```diff
@@ -389,3 +412,36 @@ func ensureMountPathCreated(targetPath string) error {
...
+		if err != nil {
+			logrus.Warnf("failed to close file: %s", err.Error())
+		}
+	}()
+	if err != nil {
+		if !os.IsExist(err) {
+			return err
+		}
+	}
+	_, err = f.Stat()
```
Comment (lpabon) [inline csi/node.go:429]: Just curious on this style. Don't want to check any stat values? May be just add some comments to why this style is being used.

File: pkg/mount/raw_mount.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	// try to find all bind mounts of raw volumes
+	if len(rawVolumeDevicesPaths) == 0 || rawVolumeDevicesPaths[0] == "" {
+		mountPointsByMajMin := make(map[string]*[]mount.Info)
+		mountPointsByTarget := make(map[string]*mount.Info)
+
+		for _, mp := range mountPoints {
+			// skip proc mount points
+			if strings.HasPrefix(mp.Mountpoint, "/proc") || mp.Root == "/null" {
```
Comment (lpabon) [inline pkg/mount/raw_mount.go:69]: Maybe add these as a const string array of locations that need to be skipped so we can add more in the future if required.

File: pkg/mount/raw_mount.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		b.trashLocation,
+	)
+	if err != nil {
+		return err
+	}
+	return b.reload(rootSubstring, newRBM.mounts[rootSubstring])
+}
+
+// this mount filtering implementation is done based on logic implemented in findmnt + libmount
+func (b *rawMounter) Load(rawVolumeDevicesPaths []string) error {
```
Comment (lpabon) [inline pkg/mount/raw_mount.go:56]: These require some unit tests

**Review by lpabon — CHANGES_REQUESTED** [review]

File: pkg/mount/raw_mount.go (hunk 4 above)
Comment (ggriffiths) [reply pkg/mount/raw_mount.go]: Will add some tests.

File: csi/node.go (hunk 2 above)
Comment (ggriffiths) [reply csi/node.go]: Removed the stat call, it was leftover from the original implementation.

File: pkg/mount/raw_mount.go (hunk 4 above)
Comment (ggriffiths) [reply pkg/mount/raw_mount.go]: Added tests in `mount_test.go` similar to other mounters @lpabon.

Comment (ggriffiths) [reply pkg/mount/raw_mount.go]: Having trouble with these tests on Travis. Working on this now.

Comment (ggriffiths) [reply pkg/mount/raw_mount.go]: Got this passing, had to switch to a test directory that was mounted in the container `/mnt/...`

**Review by lpabon — APPROVED** [review]

File: pkg/mount/raw_mount.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+		mps := *mountPointsForNumber
+		filteredMPs := []mount.Info{}
+		for i := range mps {
+			if mps[i].Mountpoint != "/dev" {
+				filteredMPs = append(filteredMPs, mps[i])
+			}
+		}
+
+		// TODO: when raw volume is mounted a few times on the same node, that case is not handled
```
Comment (adityadani) [inline pkg/mount/raw_mount.go:115]: The driver (pxd) should not allow mounting the same block device on two different mountpaths  if it does not support it.

File: pkg/mount/raw_mount.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		// skipping dev mount point
+		for _, mp := range filteredMPs {
+			devicePath := "/dev" + mp.Root
```
Comment (adityadani) [inline pkg/mount/raw_mount.go:118]: Wouldn't mp.Root contain the full device path - `/dev/sda` ?

File: pkg/mount/raw_mount.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+				rm.mounts[devicePath] = mount
+			}
+		}
+	}
+
+	// find raw volume bind mounts
+	for _, rawVolumeDevicePath := range rawVolumeDevicesPaths {
+		var mountPointForRoot *mount.Info
+		for _, mp := range mountPoints {
+			if strings.HasSuffix(rawVolumeDevicePath, mp.Root) {
```
Comment (adityadani) [inline pkg/mount/raw_mount.go:141]: The rawVolumeDevicePath should be a suffix like `/dev/pxd/pxd`. So this should change to
```
strings.HasPrefix(mp.Root, rawVolumeDevicePath)
```
Again this depends on what is mp.Root set to. I believe it is set to the whole device path

**Review by adityadani — APPROVED** [review]
Have a few comments. Happy to go over them in an offline call. Apart from that everything else looks good.

File: csi/node.go [hunk 8]
```diff
@@ -202,43 +206,44 @@ func (s *OsdCsiServer) NodeUnpublishVolume(
...
-		} else if err != nil {
-			return nil, status.Errorf(codes.NotFound, "Volume id %s not found: %s",
-				req.GetVolumeId(),
-				err.Error())
-		} else {
-			logrus.Infof("Volume %s was deleted or cannot be found", req.GetVolumeId())
-			return &csi.NodeUnpublishVolumeResponse{}, nil
-		}
+	vols, err := s.driver.Inspect([]string{volumeId})
+	if err == kvdb.ErrNotFound || len(vols) < 1 {
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply csi/node.go:224]: Good catch, this was a bad merge on my part. The raw block code was very stale.

File: pkg/mount/raw_mount.go (hunk 6 above)
Comment (ggriffiths) [reply pkg/mount/raw_mount.go]: Yes good point, I am building a test image with this change.

**Pull request merged**